support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright 2021 Snap, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <condition_variable>
+#include <mutex>
+
+namespace snapchat::djinni {
+
+template <typename T>
+struct ValueHolder {
+    using type = T;
+    std::optional<T> value;
+    T getValueUnsafe() const {return *value;}
+};
+template <>
+struct ValueHolder<void> {
+    using type = bool;
+    std::optional<bool> value;
+    void getValueUnsafe() const {}
+};
+
+template<typename T>
+struct SharedState: ValueHolder<T> {
+    std::condition_variable cv;
+    std::mutex mutex;
+    std::exception_ptr exception;
+    std::function<void(std::shared_ptr<SharedState>)> handler;
+
+    bool isReady() const {
+        return this->value.has_value() || exception != nullptr;
+    }
+};
+
+template<typename T>
+using SharedStatePtr = std::shared_ptr<SharedState<T>>;
+
+template <typename T>
+class Future;
+
+template <typename T>
+class PromiseBase {
+public:
+    Future<T> getFuture();
+
+protected:
+    void setValue(typename ValueHolder<T>::type val) {
```

Comment:
It's worth adding comment here about how this changes the internal state and the downstream effect on the handler